include/drivers/can.h

```diff
@@ -167,34 +167,44 @@ struct can_filter {
  *
  */
 struct zcan_frame {
-	/** Indicates the identifier type (standard or extended)
-	 * use can_ide enum for assignment
-	 */
-	uint32_t id_type : 1;
+	/** Message identifier*/
+	uint32_t id      : 29;
+	/** Frame is in the CAN-FD frame format */
+	uint32_t fd      : 1;
 	/** Set the message to a transmission request instead of data frame
 	 * use can_rtr enum for assignment
 	 */
 	uint32_t rtr     : 1;
-	/** Message identifier*/
-	union {
-		uint32_t std_id  : 11;
-		uint32_t ext_id  : 29;
-	};
+	/** Indicates the identifier type (standard or extended)
+	 * use can_ide enum for assignment
+	 */
+	uint32_t id_type : 1;
 	/** The length of the message (max. 8) in byte */
 	uint8_t dlc;
-	/** The message data*/
-	union {
-		uint8_t data[8];
-		uint32_t data_32[2];
-	};
+	/** Baud Rate Switch. Frame transfer with different timing during
+	 * the data phase. Only valid for CAN-FD
+	 */
+	uint8_t brs : 1;
+	/** Reserved for future flags */
+	uint8_t res : 7;
 #if defined(CONFIG_CAN_RX_TIMESTAMP)
-	/** Timer value of the CAN free running timer.
+	/** Timer value of the CAN free-running timer.
 	 * The timer is incremented every bit time and captured at the start
 	 * of frame bit (SOF).
 	 */
 	uint16_t timestamp;
+#else
+	/** @cond INTERNAL_HIDDEN */
+	uint8_t res0;  /* reserved / padding */
+	uint8_t res1;  /* reserved / padding */
+	/** @endcond */
 #endif
-} __packed;
```

Comment:
Are you sure it's safe to remove `__packed`?  Two concerns:

* Without this the placement of an instance of this type within a buffer must be naturally aligned in order for accesses to work.  Can you be sure that all instances will be aligned?
* As was pointed out to me in another context, the C standard specifies that padding can occur between members and at the end.  While a quality implementation should not introduce unnecessary padding, the standard doesn't seem to prohibit it.

I think the first point is a concern; the second might be too picky at least for common architectures.

Reply:
this frame is not sent over the wire.